server/src/routes/budget.py

```diff
@@ -146,3 +180,87 @@ def update_budget_spent(
         status_code=404,
         detail=f"Budget with month: {month} and year: {year}",
     )
+
+
+@router.post(
+    "/generateBudget",
+    response_description="Generates new budget for next month.",
+    response_model=str,
+)
+def generate_budget(keyword: str = Body(...)):
+    if keyword != "verySecurePassword":
+        raise HTTPException(
+            status_code=404,
+            detail=f"Incorrect keyword!",
+        )
+    if (all_users := user_collection.find()).count():
+        for user in all_users:
+            currentTime = datetime.now()
+            nextMonth = currentTime.month + 1
+            nextYear = currentTime.year
+            if currentTime.month == 13:
+                nextMonth = 1
+                nextYear = nextYear + 1
+
+            # Create a new budget for next month, if not there
+            if (
+                budget_collection.find_one(
+                    {"month": nextMonth, "year": nextYear, "email": user["email"]}
+                )
+            ) is None:
+                nextBudget = Budget(
+                    email=user["email"],
+                    month=nextMonth,
+                    year=nextYear,
+                    value=1000,
+                    spent=0,
+                )
+                if (
+                    budget := budget_collection.find_one(
+                        {
+                            "month": currentTime.month,
+                            "year": currentTime.year,
+                            "email": user["email"],
+                        }
+                    )
+                ) is not None:
+                    nextBudget.value = budget["value"]
+
+                nextBudget = jsonable_encoder(nextBudget)
+                new_budget = budget_collection.insert_one(nextBudget)
+
+            # Create a category budget for next month, if not there
+            # TODO: this should add budgets for all the categories in the user's expense categories list
+            #  this is because the list of categories might have changed since the last month
```

Comment:
@pochlax Added this todo item for the case where categories may have been added or deleted since last month.

In this case, I would create a category budget item for each category in the expense categories list (with each having value 0 for now)
- Then iterate over the list of categories from last month and update the budget values in the new category items created above (if it exists in the new list)
- Keep track of how much of the total month's budget (i.e. 1000) has been used up by the above
- For the amount that is left in the months budget, allocate evenly (i.e. (total remaining) / (# of new categories)) to the categories in the new list that were not updated when iterating over the old categories (i.e. newly added categories)
  - If no budget is remaining (or budget is already over), then set their values to 0

Note: use leetcode skills to do this in O(n) 😄